### ISSUE TITLE
This module should suppress errors about dc:type being empty

### ISSUE THREAD

**Issue by axfelix** [mjordan/islandora_doi_sfu#1]
Hi Mark,

Just realized one thing this implementation doesn't handle that I believe mine did -- suppressing "dc:type is empty" errors, because they're effectively being filled at POST time now.

**Comment by mjordan** [comment]
Good catch. Do you want to fix, or want me to take a stab?

**Comment by axfelix** [comment]
I can do it, just not familiar with how best to override ...

I'm assuming I need to just add a declaration to the first line of this module that replaces this array with an identical one without "type": https://github.com/mjordan/islandora_doi_framework/blob/7.x/modules/islandora_doi_datacite/includes/utilities.inc#L240 ?

Would that be namespaced somehow?

Will I inevitably break Drupal coding conventions trying to add this declaration?

**Comment by mjordan** [comment]
Would just removing 'type' from that array eliminate the problem?

**Comment by axfelix** [comment]
Yup. It needs to be there for the stock behaviour, but having this module enabled should remove it.

**Comment by mjordan** [comment]
OK, now I understand the problem. That array is defined in the base module, but your module (and possibly others, depending on what they add) need to override it. Is that an accurate characterization of the problem?

**Comment by axfelix** [comment]
Yup!

**Comment by mjordan** [comment]
OK, good catch. Lemme think about how to allow submodules to do this is a reasonably standardized way.

**Comment by mjordan** [comment]
@axfelix can you test the issue-6 branch of https://github.com/mjordan/islandora_doi_framework and the issue-1 branch of this module together? My local testing produces the error "Can't assign DOI to object rest:198 because its DC is missing one or more required values: creator, publisher, date." which is what I think you want (notice that 'type' is not in the list).

**Comment by axfelix** [comment]
Sure, will test, thanks! I'm surprised there isn't a way in Drupal to just re-declare one of the arrays you're inheriting...

**Comment by mjordan** [comment]
Implementing a `hook_alter()` is the way.... in a roundabout way.

**Comment by mjordan** [comment]
Sorry my last comment was so cryptic, and I'm not sure this expansion on it is less so. https://github.com/mjordan/islandora_doi_sfu/blob/issue-1/islandora_doi_sfu.module#L47 could just as well have been:

```php
function islandora_doi_sfu_islandora_datacite_required_values_alter(&$required_values) {
  $required_values = array('creator', 'publisher', 'date');
}
```
Reason I chose to remove 'type' instead of redefine the entire array is that if there were other modules that implemented the same hook, your change would be less destructive. However, this is largely academic since DataCite is pretty strict on its requirements.

**Comment by axfelix** [comment]
Got it. PS, this repo seems to think it has no default branch: https://help.github.com/articles/error-remote-head-refers-to-nonexistent-ref-unable-to-checkout/

**Comment by mjordan** [comment]
OK, I'll fix that ASAP.

**Comment by axfelix** [comment]
FYI, I might still make some more changes here because there are going to be several cases where we want to assign DOIs to entire collections which won't necessarily have creator / publisher / date because of how Islandora works -- I'll probably make those inherited from children in that case -- but I can probably tackle that myself.

**Comment by mjordan** [comment]
OK, can you try again?

**Comment by axfelix** [comment]
Should be fine if you toggled it on Github, I was able to workaround cloning it by specifying a branch.

**Comment by mjordan** [comment]
Want me to transfer ownership of this module to you?

**Comment by axfelix** [comment]
Nah, I can fork it and make a PR in a couple days.

**Comment by axfelix** [comment]
Hm, I'm still getting

`Notice: Undefined variable: resource_type in islandora_doi_datacite_generate_metadata() (line 283 of /var/www/html/sites/all/modules/islandora/islandora_doi_framework/modules/islandora_doi_datacite/includes/utilities.inc).`

On items that *should* be good.

**Comment by mjordan** [comment]
Ok, I'll investiage. Is that notice being dumped to the user, and if so, at what point?

**Comment by axfelix** [comment]
I get it in the DOI tab of the Manage Interface, on clicking "assign DOI."

**Comment by mjordan** [comment]
I'll need to refactor the way that https://github.com/mjordan/islandora_doi_framework/blob/issue-6/modules/islandora_doi_datacite/includes/utilities.inc#L276 works. Not sure how long/complicated.

**Comment by axfelix** [comment]
I can look into it. Sorry for making trouble, your first pass at this module was much easier for me to just fork and be done with :)

**Comment by mjordan** [comment]
No problem. To be honest, I am not happy with passing the creation of the DataCite XML off to theming. I am wondering if there should just be an additional hook that modules can implement to generate it. That way, you'd have full control. Passing in the DOI assingment form values might be all you need. Not sure, let me think about this a bit more.

**Comment by mjordan** [comment]
@axfelix I have some time this week to spend on this. Any additional/new use cases I should know about?

**Comment by axfelix** [comment]
Nothing major. If you could look into the existing issue in this thread that'd be great.
